gcf_data_mapper/parsers/document.py

```diff
@@ -1,19 +1,161 @@
+    return False
+
+
+def contains_invalid_paths(urls: list[str]) -> bool:
+    """
+    Checks a list of urls for any malformed entries
+
+    param: list[str] urls: A list of urls
+    return bool: Returns true if malformed urls are present, or false if not
+    """
+    for url in urls:
+        parsed_url = urlparse(url)
+        path = parsed_url.path
+        # Reserved and unreserved characters per RFC 3986
+        reserved_and_unreserved_characters = ":/?#[]@!$&'()*+,;=-_.~"
+        if any(
+            not (c.isalnum() or c in reserved_and_unreserved_characters) for c in path
+        ):
+            return True
+    return False
+
+
+def validate_urls(urls: list[str], doc_id: str) -> None:
+    """
+    Validates a list of URLs for empty, duplicate, and malformed entries.
+
+    param: list[str] urls : A list of urls
+    param: str doc_id: The document id of the invalid source url/s
+    raises ValueError: If the list contains duplicate, empty or malformed url/s
+    """
+    if contains_empty_urls(urls):
+        raise ValueError(
+            f"Empty URL found in list of translated urls. DocumentId : {doc_id}"
+        )
+    if contains_duplicate_urls(urls):
+        raise ValueError(
+            f"Duplicate URLs found in list of translated urls. DocumentId : {doc_id}"
+        )
+    if contains_invalid_paths(urls):
+        raise ValueError(
+            f"Malformed url found in list of translated urls. DocumentId : {doc_id}"
+        )
+
+
+def map_translated_files(translated_files_row: pd.Series) -> list[dict]:
+    """
+    Maps the GCF document with translated versions into the new json structure
+
+    :param pd.Series translated_files_row: A row from the DataFrame containing the 'Translated files' field, which holds a string of translated source URLs separated by the pipe (|) character. This string includes multiple URLs for translated documents in various languages.
+    :return: A list of mcf document objects, each with a different source url reflecting the translated version of the original document
+    """
+
+    mapped_documents = []
+
+    concatenated_string_of_url_docs = str(
+        translated_files_row[RequiredColumns.TRANSLATED_FILES.value]
+    )
+    url_docs = concatenated_string_of_url_docs.split("|")
+
+    doc_id = translated_files_row.iloc[0]
+
+    try:
+        validate_urls(url_docs, doc_id)
+        for url in url_docs:
+            mapped_documents.append(
+                {
+                    "metadata": {
+                        "type": translated_files_row[RequiredColumns.TYPE.value]
+                    },
+                    "title": translated_files_row[RequiredColumns.TITLE.value],
+                    "source_url": url.strip(),
+                    "variant_name": "Translated",
+                }
+            )
+        return mapped_documents
+    except Exception as e:
+        raise e
+
+
 def document(mcf_docs: pd.DataFrame, debug: bool) -> list[Optional[dict[str, Any]]]:
```

Comment:
Slightly pedantic - but remember despite the name of the file, the docs file we are working with is for GCF documents, not all of the documents from all of the 4 MCFs. Can we rename this parameter to make that more explicit please?